services/rag_service/embedding_service.py

```diff
@@ -0,0 +1,125 @@
+        while q:
+            node = q.popleft()
+            doc = self.data_service.get_user_documentation(user_id, node.id)
+            await self.generate_markdown_embeddings_for_doc(doc, repo_id)
+
+            for child in node.children:
+                if child.completion_status == StatusEnum.COMPLETED:
+                    q.append(child)
+    
+    async def generate_markdown_embeddings_for_doc(self, doc: FirestoreDoc, repo_id: str):
+        markdown = doc.markdown_content
+        markdown_regex = [
+            # Note that this splitter doesn't handle horizontal lines defined
+            # by *three or more* of ***, ---, or ___, but this is not handled
+            "",
+            " ",
+            "\n",
+            "\n\n",
+            # Horizontal lines
+            "\n___+\n",
+            "\n---+\n",
+            "\n\\*\\*\\*+\n",
+            # Note the alternative syntax for headings (below) is not handled here
+            # Heading level 2
+            # ---------------
+            # End of code block
+            "```\n",
+            # First, try to split along Markdown headings (starting with level 2)
+            "\n#{1,6} "
+        ]
+
+        # There is a max of 2048 embeddings that can be generated at a time, so we need to "group" the chunks
+        chunks = self.text_chunker.chunk(markdown, markdown_regex)
+        grouped_chunks = [chunks[i:i+2048] for i in range(0, len(chunks), 2048)]
+
+        chunk_index = 0
+        for list_of_chunks in grouped_chunks:
+            embeddings = await self.embedding_client.generate_embedding(
+                model=EmbeddingModelEnum.BGE_LARGE,
+                input=list_of_chunks
+            )
+            # The embeddings.data can have up to 2048 embeddings, upserting individually is slow
+            # Pinecone has a limit of 100 upserts per request, so we will "group" the vectors into groups of 100
+            vectors = []
+            n = len(embeddings.data)
+            for i, embedding in enumerate(embeddings.data):
+                vectors.append({
+                    "id": doc.id + f"-{chunk_index}",
+                    "values": embedding.embedding,
+                    "metadata": {
+                        "chunk_content": list_of_chunks[embedding.index],
+                        "doc_id": doc.id,
+                    }
+                })
+                if len(vectors) == 100 or i == n - 1:
+                    self.vector_database_client.upsert(vectors, repo_id)
+                    vectors = []
+                chunk_index += 1
+                
+
+
+if __name__ == "__main__":
+    load_dotenv()
+    firebase_app = firebase_admin.initialize_app(
+        credential=None,
+        options={"storageBucket": os.getenv("CLOUD_STORAGE_BUCKET")}
+    )
+    anyscale_client = AnyscaleClient(api_key=os.getenv("ANYSCALE_API_KEY"))
+    vector_database_client = PineconeClient(api_key=os.getenv("PINECONE_API_KEY"), index="rocketdocs-repos-1")
```

Comment:
**suggestion:** please add a factory function for this service. You can look at any of our other classes for an example, usually starts with `get_class_name()`